sdk/metrics/src/jmhBasedTest/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorageGarbageCollectionBenchmarkTest.java

```diff
@@ -0,0 +1,97 @@
+   *
+   * <p>It runs the JMH test {@link AsynchronousMetricStorageGarbageCollectionBenchmark} with GC
+   * profiler, and measures for each parameter combination the garbage collector normalized rate
+   * (bytes allocated per Operation).
+   *
+   * <p>Memory allocations can be hidden even at an innocent foreach loop on a collection, which
+   * under the hood allocates an internal object O(N) times. Someone can accidentally refactor such
+   * loop, resulting in 30% increase of garbage collected objects during a single collect() run.
+   */
+  @SuppressWarnings("rawtypes")
+  @Test
+  public void normalizedAllocationRateTest() throws RunnerException {
+    // Runs AsynchronousMetricStorageMemoryProfilingBenchmark
+    // with garbage collection profiler
+    Options opt =
+        new OptionsBuilder()
+            .include(AsynchronousMetricStorageGarbageCollectionBenchmark.class.getSimpleName())
+            .addProfiler("gc")
+            .shouldFailOnError(true)
+            .build();
+    Collection<RunResult> results = new Runner(opt).run();
+
+    // Collect the normalized GC allocation rate per parameters combination
+    Map<String, Map<String, Double>> resultMap = new HashMap<>();
+    for (RunResult result : results) {
+      for (BenchmarkResult benchmarkResult : result.getBenchmarkResults()) {
+        BenchmarkParams benchmarkParams = benchmarkResult.getParams();
+
+        String memoryMode = benchmarkParams.getParam("memoryMode");
+        String aggregationTemporality = benchmarkParams.getParam("aggregationTemporality");
+        assertThat(memoryMode).isNotNull();
+        assertThat(aggregationTemporality).isNotNull();
+
+        Map<String, Result> secondaryResults = benchmarkResult.getSecondaryResults();
+        Result allocRateNorm = secondaryResults.get("gc.alloc.rate.norm");
+        assertThat(allocRateNorm)
+            .describedAs("Allocation rate in secondary results: %s", secondaryResults)
+            .isNotNull();
+
+        resultMap
+            .computeIfAbsent(aggregationTemporality, k -> new HashMap<>())
+            .put(memoryMode, allocRateNorm.getScore());
+      }
+    }
+
+    assertThat(resultMap).hasSameSizeAs(AggregationTemporality.values());
+
+    // Asserts that reusable data GC allocation rate is a tiny fraction of immutable data
+    // GC allocation rate
+    resultMap.forEach(
+        (aggregationTemporality, memoryModeToAllocRateMap) -> {
+          Double immutableDataAllocRate =
+              memoryModeToAllocRateMap.get(MemoryMode.IMMUTABLE_DATA.toString());
+          Double reusableDataAllocRate =
+              memoryModeToAllocRateMap.get(MemoryMode.REUSABLE_DATA.toString());
+
+          assertThat(immutableDataAllocRate).isNotNull().isNotZero();
+          assertThat(reusableDataAllocRate).isNotNull().isNotZero();
+          assertThat(100 - (reusableDataAllocRate / immutableDataAllocRate) * 100)
+              .isCloseTo(99.8, Offset.offset(2.0));
```

Comment:
I have mixed feelings about running this by default in the test suite. On one hand, it helps avoid accidental changes that reduce performance in surprising ways. On the other hand, this test increases the time to run `./gradlew :sdk:metrics:build` on my machine by > 50%, from ~22s to ~34s. (See this [build scan](https://ge.opentelemetry.io/s/fq5s7hpeq2pne/tests/overview?toggled=WyI6c2RrOm1ldHJpY3M6dGVzdC0wIl0) for details).

What if we run this test only in the github actions build, where the extra time isn't as cumbersome and where we still get protection against performance reducing changes. Can use the `CI=true` environment variable which is [set by default in github actions](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables).

Reply:
I can do this, inside the only single method of this test class:

```
  @Test
  public void normalizedAllocationRateTest() throws RunnerException {
    Assumptions.assumeTrue(
        "true".equals(System.getenv("CI")),
        "This test should only run in GitHub CI since it's long");

```

The test will only run if CI == "true", or skipped otherwise (local workstation).
I cane make something much more robust, but given this is the only test so far with lengthy runtime, it can be enough or now.
WDYT?